### PR TITLE
[Versioning] Rebase to OpenSearch version 1.0.0

### LINKED ISSUE #417: [Version] Rebase versioning to 1.0.0 with bwc support
This issue documents the rebase of the versioning from the forked predecessor version 7.10.2* to OpenSearch 1.0.0.

### Existing Version Implementation

Versioning is essentially implemented as a class wrapper around a version id integer that provides a semantic API for relating between versions. The logic for the version id integer is XXYYZZAA where:

    * XX is major version
    * YY is minor version
    * ZZ is revision / bugfix
    * AA is alpha/beta/RC/GA builds  (values < 25 are alpha, 25 >= and < 50 are beta, 50 >= and <99 are RC, and 99 is GA)

This semantically enables before & after checks on the version id integers. Following this contract the maximum possible version number is 99.99.99.99 (or 99999999 in integer space).

### Rebased OpenSearch Version Change

To keep backwards compatibility with the predecessor (e.g., 7.10 of the predecessor is treated as an earlier version of OpenSearch 1.0), the following must hold true:

    * OpenSearch versionid integer space must be higher than the predecessors integer space. 
    * Future releases of the predecessor must remain separated and easily identifiable from future releases of OpenSearch

The initial approach is as follows:

1. split the existing `Version` class into two:
    a. new `LegacyESVersion` class for holding 6.x and 7.x version ids and semantics
    b. existing `Version` for holding the rebased OpenSearch version 1.x version ids and semantics (e.g., OpenSearch 1.0.0.alpha1 == Version.V_1_0_0_alpha1).
2. allocate a new integer space for OpenSearch versioning
    a. Legacy versionid space is a maximum of 99999999; mapping to 27 bits
    b. OpenSearchID space will use the 28th bit as an identifier; when flipped to `1` the versionid is recognized as an OpenSearch version id.
    c. The remaining version id logic of the predecessor is retained for versions of OpenSearch (e.g., XX.YY.ZZ.AA). This keeps all of the semantics throughout the code base unchanged.
3. `LegacyESVersion` is used in all existing version check locations. (e.g., `LegacyESVersion.V_7_10_0`)
4. Breaking changes or new features that are not backward compatible will use the existing `Version` class like before (but this will now represent OpenSearch versions going forward) 

### Examples

1. The following code executes for indexes created on or later than Predecessor version 6.6.0

```java
 if (indexCreatedVersion.onOrAfter(LegacyESVersion.V_6_6_0) {
    // do some stuff
 }
```

2. The following code executes for transport data received by a client deployed before the OpenSearch 1.0.0 alpha 1 release (forward compatibility)

```java
 if (streamInput.getVersion().onOrBefore(Version.V_1_0_0_alpha1)) {
    // do some stuff for a legacy client
 } else {
    // do some new cool stuff
 }
```

### Notes

**logs**
./gradlew bwcTest
> Task :distribution:bwc:bugfix:checkoutBwcBranch FAILED
Performing checkout of opensearch-project/7.10...
Exec output and error:
| Output for git:error: pathspec 'opensearch-project/7.10' did not match any file(s) known to git

> Task :server:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:checkoutBwcBranch'.
> Process 'command 'git'' finished with non-zero exit value 1

**Subtasks**
- [ ] :distribution:bwc:bugfix:checkoutBwcBranch FAILED

* This approach is implemented on a feature branch and will be merged to `main` either for, or initially following, the initial opening of the repositories, and this issue will remain open as a discuss issue for anyone that wants to improve or propose a new or complete change of the design. (encouraged)
* On day 1 of the opening of the repository Initial integration and unit tests may or may not be completely finished.

### PR DESCRIPTION
**Pull request by nknize**
### Description
This PR rebases the versioning framework to OpenSearch version 1.0.0 while maintaining backwards compatibility with legacy versions from the point in time of the code fork. Note that this is also (technically) forward compatible with future versions of the legacy versioning ids. A followup PR should include graceful deprecation logging for any legacy version greater than 7.10.2.
 
### Issues Resolved
resolves #417
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

File: buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java [hunk 1]
```diff
@@ -306,8 +314,13 @@ private Version getLatestVersionByKey(Map<Integer, List<Version>> groupByMajor,
...
 
     private Map<Integer, List<Version>> getReleasedMajorGroupedByMinor() {
-        List<Version> currentMajorVersions = groupByMajor.get(currentVersion.getMajor());
-        List<Version> previousMajorVersions = groupByMajor.get(currentVersion.getMajor() - 1);
+        int currentMajor = currentVersion.getMajor();
+        List<Version> currentMajorVersions = groupByMajor.get(currentMajor);
+        List<Version> previousMajorVersions = groupByMajor.get(currentMajor == 1 ? 7 : currentVersion.getMajor() - 1);
+        if (previousMajorVersions == null) {
+            System.out.println("size: " + groupByMajor.size());
+            System.out.println(currentVersion.getMajor());
```
Comment (setiah) [inline buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java:322]: Are these prints lying around from local debugging?

Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java]: yes.. thx! removed.

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
start gradle check

**Comment by peterzhuamazon** [comment]
start gradle check

**Review by saratvemulapalli — APPROVED** [review]
Phew that was a big one.
The changes look good to me. Just a couple of minor comments.

Thanks a lot for the change!

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java [hunk 2]
```diff
@@ -1151,13 +1151,13 @@ private void createConfiguration() {
...
         }
-        if (getVersion().getMajor() >= 6) {
+        if (getVersion().onOrAfter("6.0.0")) {
             baseConfig.put("cluster.routing.allocation.disk.watermark.flood_stage", "1b");
         }
         // Temporarily disable the real memory usage circuit breaker. It depends on real memory usage which we have no full control
         // over and the REST client will not retry on circuit breaking exceptions yet (see #31986 for details). Once the REST client
         // can retry on circuit breaking exceptions, we can revert again to the default configuration.
-        if (getVersion().getMajor() >= 7) {
+        if (getVersion().onOrAfter("7.0.0")) {
```
Comment (setiah) [inline buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:1160]: nit:
```suggestion
        if (getVersion().onOrAfter(LegacyESVersion.V_7_0_0)) {
```

**Comment by nknize** [comment]
start gradle check

File: buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java (hunk 2 above)
Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/testclusters/OpenSearchNode.java:1160]: build-tools has no dependency on the `server` module so `LegacyESVersion` is not available in the `buildSrc` module.

File: buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java [hunk 3]
```diff
@@ -144,12 +144,17 @@ public BwcVersions(SortedSet<Version> allVersions, Version currentVersionPropert
             throw new IllegalArgumentException("Could not parse any versions");
         }
 
+        // hack: this is horribly volatile like this entire logic; fix
         currentVersion = allVersions.last();
 
         groupByMajor = allVersions.stream()
             // We only care about the last 2 majors when it comes to BWC.
```
[the remark this answers is not in the document]
Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java:151]: :+1:

File: buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java [hunk 4]
```diff
@@ -353,23 +362,36 @@ public void compareToAuthoritative(List<Version> authoritativeReleasedVersions)
...
 
     public List<Version> getIndexCompatible() {
-        return unmodifiableList(
-            Stream.concat(groupByMajor.get(currentVersion.getMajor() - 1).stream(), groupByMajor.get(currentVersion.getMajor()).stream())
-                .filter(version -> version.equals(currentVersion) == false)
-                .collect(Collectors.toList())
-        );
-
+        int currentMajor = currentVersion.getMajor();
+        int prevMajor = currentMajor == 1 ? 7 : currentMajor - 1;
```
[the remark this answers is not in the document]
Comment (nknize) [reply buildSrc/src/main/java/org/opensearch/gradle/BwcVersions.java:366]: :+1:

File: server/src/main/java/org/opensearch/Version.java [hunk 5]
```diff
@@ -363,7 +285,14 @@ private static Version fromStringSlow(String version) {
     public final org.apache.lucene.util.Version luceneVersion;
 
     Version(int id, org.apache.lucene.util.Version luceneVersion) {
-        this.id = id;
+        // flip the 28th bit of the ID; identify as an opensearch vs legacy system:
+        // we start from version 1 for opensearch, so ignore the 0 (empty) version
+        if(id != 0) {
+            this.id = id ^ MASK;
+            id &= 0xF7FFFFFF;
+        } else {
```
Comment (setiah) [inline server/src/main/java/org/opensearch/Version.java:293]: should there be an else case?

**Review by setiah — COMMENTED** [review]
Looks good to me. Thanks for the huge effort. 

Not sure if I didn't see this or missed, are there any UTs for specifically checking Opensearch v1.0.0 bwc behavior with legacy versions?

File: server/src/main/java/org/opensearch/Version.java (hunk 5 above)
Comment (nknize) [reply server/src/main/java/org/opensearch/Version.java:293]: It's to support `V_EMPTY` but it looks like that's not even used anywhere. So we could probably remove it. I'd prefer doing that in a separate PR?

**Comment by nknize** [comment]
start gradle check

File: server/src/main/java/org/opensearch/Version.java (hunk 5 above)
Comment (setiah) [reply server/src/main/java/org/opensearch/Version.java:293]: Sounds good. Thanks!

**Review by setiah — APPROVED** [review]

File: server/src/main/java/org/opensearch/Version.java (hunk 5 above)
Comment (nknize) [reply server/src/main/java/org/opensearch/Version.java:293]: And good catch! I'll open an issue.

**Pull request merged**